File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/NetworkDebugInfoWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.xmtp.android.library.ApiStats
import org.xmtp.android.library.IdentityStats
import org.xmtp.android.library.XMTPDebugInformation

class NetworkDebugInfoWrapper {
    companion object {
        val gson: Gson = GsonBuilder().create()
        private fun encodeToObj(info: XMTPDebugInformation): Map<String, Any> {
            return mapOf(
                "apiStatistics" to ApiStatsWrapper.encode(info.apiStatistics),
                "identityStatistics" to IdentityStatsWrapper.encode(info.identityStatistics),
                "aggregateStatistics" to info.aggregateStatistics,
            )
        }

        fun encode(info: XMTPDebugInformation): String {
            val obj = encodeToObj(info)
            return gson.toJson(obj)
        }
    }
}

private class ApiStatsWrapper {
    companion object {
        val gson: Gson = GsonBuilder().create()
        private fun encodeToObj(info: ApiStats): Map<String, Any> {
            return mapOf(
                "uploadKeyPackage" to info.uploadKeyPackage,
                "fetchKeyPackage" to info.fetchKeyPackage,
                "sendGroupMessages" to info.sendGroupMessages,
                "sendWelcomeMessages" to info.sendWelcomeMessages,
                "queryGroupMessages" to info.queryGroupMessages,
                "queryWelcomeMessages" to info.queryWelcomeMessages,
                "subscribeMessages" to info.subscribeMessages,
                "subscribeWelcomes" to info.subscribeWelcomes,
            )
        }

        fun encode(info: ApiStats): String {
            val obj = encodeToObj(info)
            return gson.toJson(obj)
        }
    }
}

private class IdentityStatsWrapper {
    companion object {
        val gson: Gson = GsonBuilder().create()
        private fun encodeToObj(info: IdentityStats): Map<String, Any> {
            return mapOf(
                "publishIdentityUpdate" to info.publishIdentityUpdate,
                "getIdentityUpdatesV2" to info.getIdentityUpdatesV2,
                "getInboxIds" to info.getInboxIds,
                "verifySmartContractWalletSignature" to info.verifySmartContractWalletSignature,
            )
        }

        fun encode(info: IdentityStats): String {
            val obj = encodeToObj(info)
            return gson.toJson(obj)
        }
    }
}

File: ios/Wrappers/NetworkDebugInfoWrapper.swift
//
//  NetworkDebugInfoWrapper.swift
//  Pods
//
//  Created by Naomi Plasterer on 5/23/25.
//

import Foundation
import XMTP


struct ApiStatsWrapper {
	static func encodeToObj(_ info: ApiStats) -> [String: Any] {
		return [
			"uploadKeyPackage": info.uploadKeyPackage,
			"fetchKeyPackage": info.fetchKeyPackage,
			"sendGroupMessages": info.sendGroupMessages,
			"sendWelcomeMessages": info.sendWelcomeMessages,
			"queryGroupMessages": info.queryGroupMessages,
			"queryWelcomeMessages": info.queryWelcomeMessages,
			"subscribeMessages": info.subscribeMessages,
			"subscribeWelcomes": info.subscribeWelcomes
		]
	}

	static func encode(_ info: ApiStats) throws -> String {
		let obj = encodeToObj(info)
		let data = try JSONSerialization.data(withJSONObject: obj)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode API stats")
		}
		return result
	}
}


struct IdentityStatsWrapper {
	static func encodeToObj(_ info: IdentityStats) -> [String: Any] {
		return [
			"publishIdentityUpdate": info.publishIdentityUpdate,
			"getIdentityUpdatesV2": info.getIdentityUpdatesV2,
			"getInboxIds": info.getInboxIds,
			"verifySmartContractWalletSignature": info.verifySmartContractWalletSignature
		]
	}

	static func encode(_ info: IdentityStats) throws -> String {
		let obj = encodeToObj(info)
		let data = try JSONSerialization.data(withJSONObject: obj)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode identity stats")
		}
		return result
	}
}


struct NetworkDebugInfoWrapper {
	static func encodeToObj(_ info: XMTPDebugInformation) -> [String: Any] {
		return [
			"apiStatistics": ApiStatsWrapper.encodeToObj(info.apiStatistics),
			"identityStatistics": IdentityStatsWrapper.encodeToObj(info.identityStatistics),
			"aggregateStatistics": info.aggregateStatistics
		]
	}

	static func encode(_ info: XMTPDebugInformation) throws -> String {
		let obj = encodeToObj(info)
		let data = try JSONSerialization.data(withJSONObject: obj)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode network debug info")
		}
		return result
	}
}
